src/aws/mod.rs

```diff
@@ -54,113 +66,269 @@ impl ModerationProvider for Rekognition {
-                match hyper::body::to_bytes(response.into_body()).await {
-                    Ok(bytes) => match serde_json::from_slice::<RekognitionResponse>(&bytes) {
-                        Ok(r) => Ok(r),
-                        Err(e) => Err(Box::new(e)),
-                    },
-                    Err(e) => Err(Box::new(e)),
+        document: &Document,
+        context: Arc<Context>,
+    ) -> Result<Option<RekognitionResponse>, AwsError> {
+        let bytes: &hyper::body::Bytes = &document.bytes;
+        if document.is_video() {
+            if self.video_config.is_none() {
+                Err("aws video config missing".into())
+            } else {
+                // set up region a shared config
+                let bucket = &self.video_config.as_ref().unwrap().bucket;
+                let sns_topic_arn = &self.video_config.as_ref().unwrap().sns_topic_arn;
+                let s3_permits = &self.video_config.as_ref().unwrap().s3_jobs;
+
+                let role_arn = &self.video_config.as_ref().unwrap().role_arn;
+
+                let region = aws_sdk_s3::Region::new(self.region.as_str().to_owned());
+                //we need this future to setup client ,
+                //hence we can't set up the client in new ,as new is
+                //not async
+
+                let shared_config = aws_config::from_env().region(region).load().await;
+
+                use crate::aws::s3::S3;
+                let s3 = S3::new(&self.region.as_str().to_owned(), bucket, s3_permits).await;
+
+                s3.add_to_bucket(document).await?;
+
+                let channel = NotificationChannel::builder()
+                    .sns_topic_arn(sns_topic_arn)
+                    .role_arn(role_arn)
+                    .build();
+
+                // set up rekognition client
+                let client_rekognition = ClientRekognition::new(&shared_config);
+
+                // start video moderation and get job id
+                let job_id = self
+                    .start_video_moderation(&client_rekognition, channel, bucket, &document.url)
+                    .await?;
+
+                if let Err(e) = context
+                    .database
+                    .add_job(&job_id, &document.url, &ModerationStatus::Pending)
+                    .await
+                {
+                    error!("{}", e);
+                }
+                // get reuslts
+                let reko_permit = self.rekognition_permits.clone();
+                let ctx = context.clone();
+                let id = document.id;
+                let url = document.url.clone();
+
+                tokio::task::spawn(async move {
+                    match Rekognition::get_moderation_results(
+                        &client_rekognition,
+                        job_id.clone(),
+                        reko_permit,
+                    )
+                    .await
+                    {
+                        Ok(r) => {
```

Comment:
Instead of doing these updates in the spawned thread, have master thread that iterates through the job table and updates accordingly